### ISSUE TITLE
How can I use the class `MetaLayer` to implement an attention mechanism?

### ISSUE THREAD

**Issue by dmeoli** [pyg-team/pytorch_geometric#3223]
I'm trying to extend a project which implements a GNN using Battaglia et al.'s definition through the `MetaLayer` class. 
I would like to include some attention mechanisms as defined [here](https://github.com/deepmind/graph_nets/blob/64771dff0d74ca8e77b1f1dcd5a7d26634356d61/graph_nets/modules.py#L657), so how can I use the `MetaLayer` class to implement this TF code?

**Comment by rusty1s** [comment]
IMO, the easiest way to implement this is to make use of the `MessagePassing` interface rather than the `MetaLayer`, in which you return both transformed node features and edge features. Here is a minimal example to allow for attention based on both node features and edge features: https://github.com/pyg-team/pytorch_geometric/discussions/3209#discussioncomment-1373286

If you still wanna use the `MetaLayer`, then you need to update the `NodeModel` to compute attention coefficients and apply normalization based on it, e.g.:
```python
from torch_geometric.utils import softmax

class NodeModel(torch.nn.Module):
    def __init__(self):
        super(NodeModel, self).__init__()
        self.att_mlp = # Maps to number_of_heads
        self.transform_mlp = # Maps to number of heads * number of output channels

    def forward(self, x, edge_index, edge_attr, u, batch):
        row, col = edge_index
        out = torch.cat([x[row], edge_attr, x[col]], dim=1)
        att = self.att_mlp(out)
        att = softmax(att, col)  # Normalize across local neighborhoods
        out = self.transform_mlp(out)
        out = out.view(-1, num_heads, num_features) * att.view(-1, num_heads, 1)
        out = out.view(-1, num_heads * num_features)
        out = scatter_add(out, col, dim=0, dim_size=x.size(0))
        return out
```

**Comment by dmeoli** [comment]
Actually, my code is: 

`
class ModifiedMetaLayer(MetaLayer):

    def forward(
            self, x, edge_index, edge_attr=None, u=None, v_indices=None, e_indices=None
    ):
        row, col = edge_index

        if self.edge_model is not None:
            edge_attr = self.edge_model(x[row], x[col], edge_attr, u, e_indices)

        if self.node_model is not None:
            x = self.node_model(x, edge_index, edge_attr, u, v_indices)

        if self.global_model is not None:
            u = self.global_model(x, edge_attr, u, v_indices, e_indices)

        return x, edge_attr, u

def get_mlp(
        in_size,
        out_size,
        n_hidden,
        hidden_size,
        activation=ReLU,
        activate_last=True,
        layer_norm=True
):
    arch = []
    l_in = in_size
    for l_idx in range(n_hidden):
        arch.append(Lin(l_in, hidden_size))
        arch.append(activation())
        l_in = hidden_size

    arch.append(Lin(l_in, out_size))

    if activate_last:
        arch.append(activation())

        if layer_norm:
            arch.append(LayerNorm(out_size))

    return Seq(*arch)


class GraphNet:

    def __init__(
            self,
            in_dims,
            out_dims,
            independent=False,
            e2v_agg="sum",
            n_hidden=1,
            hidden_size=64,
            activation=ReLU,
            layer_norm=True
    ):
        if e2v_agg not in ["sum", "mean"]:
            raise ValueError("Unknown aggregation function.")

        v_in = in_dims[0]  # n_node_features_in
        e_in = in_dims[1]  # n_edge_features_in
        u_in = in_dims[2]  # n_global_features_in

        v_out = out_dims[0]  # n_node_features_out
        e_out = out_dims[1]  # n_edge_features_out
        u_out = out_dims[2]  # n_global_features_out

        class EdgeModel(torch.nn.Module):
            def __init__(self):
                super(EdgeModel, self).__init__()

                if independent:
                    self.edge_mlp = get_mlp(
                        e_in,  # n_edge_features_in
                        e_out,  # n_edge_features_out
                        n_hidden,
                        hidden_size,
                        activation=activation,
                        layer_norm=layer_norm
                    )
                else:
                    self.edge_mlp = get_mlp(
                        e_in + 2 * v_in + u_in,  # n_edge_features_in + 2 * n_node_features_in + n_global_features_in
                        e_out,
                        n_hidden,
                        hidden_size,
                        activation=activation,
                        layer_norm=layer_norm
                    )

            def forward(self, src, target, edge_attr, u=None, e_indices=None):
                # src, target: [E, F_x], where E is the number of edges.
                # edge_attr: [E, F_e]
                # u: [B, F_u], where B is the number of graphs.
                # e_indices: [E] with max entry B - 1.
                if independent:
                    return self.edge_mlp(edge_attr)

                out = torch.cat([src, target, edge_attr, u[e_indices]], 1)
                return self.edge_mlp(out)

        class NodeModel(torch.nn.Module):
            def __init__(self):
                super(NodeModel, self).__init__()

                if independent:
                    self.node_mlp = get_mlp(
                        v_in,  # n_node_features_in
                        v_out,  # n_node_features_out
                        n_hidden,
                        hidden_size,
                        activation=activation,
                        layer_norm=layer_norm
                    )
                else:
                    self.node_mlp = get_mlp(
                        v_in + e_out + u_in,  # n_node_features_in + n_edge_features_out + n_global_features_in
                        v_out,  # n_node_features_out
                        n_hidden,
                        hidden_size,
                        activation=activation,
                        layer_norm=layer_norm
                    )

            def forward(self, x, edge_index, edge_attr, u=None, v_indices=None):
                # x: [N, F_x], where N is the number of nodes.
                # edge_index: [2, E] with max entry N - 1.
                # edge_attr: [E, F_e]
                # u: [B, F_u], where B is the number of graphs.
                # v_indices: [N] with max entry B - 1.
                if independent:
                    return self.node_mlp(x)

                row, col = edge_index
                if e2v_agg == "sum":
                    out = scatter_add(edge_attr, row, dim=0, dim_size=x.size(0))
                elif e2v_agg == "mean":
                    out = scatter_mean(edge_attr, row, dim=0, dim_size=x.size(0))
                out = torch.cat([x, out, u[v_indices]], dim=1)
                return self.node_mlp(out)

        class GlobalModel(torch.nn.Module):

            def __init__(self):
                super(GlobalModel, self).__init__()

                if independent:
                    self.global_mlp = get_mlp(
                        u_in,  # n_global_features_in
                        u_out,  # n_global_features_out
                        n_hidden,
                        hidden_size,
                        activation=activation,
                        layer_norm=layer_norm
                    )
                else:
                    self.global_mlp = get_mlp(
                        u_in + v_out + e_out,  # n_global_features_in + n_node_features_out + n_edge_features_out
                        u_out,  # n_global_features_out
                        n_hidden,
                        hidden_size,
                        activation=activation,
                        layer_norm=layer_norm
                    )

            def forward(self, x, edge_attr, u, v_indices, e_indices):
                # x: [N, F_x], where N is the number of nodes.
                # edge_attr: [E, F_e]
                # u: [B, F_u], where B is the number of graphs.
                # e_indices: [E] with max entry B - 1.
                if independent:
                    return self.global_mlp(u)

                out = torch.cat([u,
                                 scatter_mean(x, v_indices, dim=0),
                                 scatter_mean(edge_attr, e_indices, dim=0)], dim=1)
                return self.global_mlp(out)

        self.op = ModifiedMetaLayer(EdgeModel(), NodeModel(), GlobalModel())

    def forward(
            self, x, edge_index, edge_attr=None, u=None, v_indices=None, e_indices=None
    ):
        return self.op(x, edge_index, edge_attr, u, v_indices, e_indices)
`

Then, the `GraphNet` class is used to build an `EncoderCoreDecoder` architecture as specified in Battaglia et al..

I have both node and edge features, so (according to this implementation) I need to change the `NodeModel` and `EdgeModel` classes to compute attention coefficients.

But how can I rewrite this class using the `MessagePassing` interface rather than the `MetaLayer` in order to add some extra attention layer (e.g., `GATConv`)?

Thx

**Comment by rusty1s** [comment]
The  `MessagePassing` class helps you to perform message passing to compute new node features based on neighboring ones. You can easily extend it to incorporate updating edge representations as well:
```
class GraphNet(MessagePassing):
    def __init__(self, ...)
        super().__init__(self, aggr='mean')
        self.node_MLP = ...
        self.edge_MLP = ...
        
    def forward(self, x, edge_index, edge_attr):
         row, col = edge_index
         edge_attr = self.edge_MLP(torch.cat([x[row], x[col], edge_attr], dim=-1)
         x = self.propagate(edge_index, x=x, edge_attr=edge_attr)
         return x, edge_attr
     
    def message(self, x_i, x_j, edge_attr):
        edge_attr = self.node_MLP(torch.cat([x_i, x_j, edge_attr], dim=-1)
```
